### ISSUE TITLE
ncdirect_printf_aligned also returns number of characters printed

### ISSUE THREAD

**Issue by igo95862** [dankamongmen/notcurses#1074]

**Comment by igo95862** [comment]
Also it seems like using NCALIGN_UNALIGNED will cause y parameter to have no effect.

**Comment by dankamongmen** [comment]
It also looks to be broken for multicell glyphs, taking a look at it =[


`  const size_t len = strlen(r);`


well, this function is no good!

**Comment by igo95862** [comment]
Does that mean the ncdirect_putstr should only be used for ncdirect

**Comment by igo95862** [comment]
The issue with ncdirect_putstr it takes a channels argument meaning the whatever the values of default background and foreground colors are they are not used.

**Comment by dankamongmen** [comment]
Anything prefixed with `ncdirect_` ought only be used with `ncdirect`, yes.

What I said about multicell glyphs refers to calculating the width using `strlen()`. `strlen()` counts bytes, not columns, which is what we want for aligning output. I need change it to use `ncstrwidth()`.

**Comment by dankamongmen** [comment]
(i almost never use direct mode, so it tends to be a bit rough, sorry about that)

**Comment by dankamongmen** [comment]
Hrmmm, I can't reproduce this behavior you speak of where `NCALIGN_NONE` results in an ignored `y`...

**Comment by dankamongmen** [comment]
> The issue with ncdirect_putstr it takes a channels argument meaning the whatever the values of default background and foreground colors are they are not used.

I'm not sure I understand what you mean here. When I call `ncdirect_putstr()`, it absolutely sets the colors to those passed in. That's honestly the main reason it exists. Otherwise you could just `fprintf()` directly.

**Comment by igo95862** [comment]
https://github.com/dankamongmen/notcurses/blob/19efd2c10aaa556537b37266bf1783b713ea7dd8/include/notcurses/direct.h#L52

> (struct ncdirect* nc, uint64_t channels, const char* utf8)

**Comment by dankamongmen** [comment]
Passing 0 ought make it use the terminal's default foreground and background color (for me, this is white on black).

**Comment by dankamongmen** [comment]
If you want e.g. black (RGB 0), you want something like:

`uint64_t channels = CHANNELS_RGB_INITIALIZER(0, 0, 0, 0xff, 0, 0);`

and pass that `channels` into `ncdirect_putstr()`. You can also use  `ncdirect_fg_rgb()` etc.

**Comment by igo95862** [comment]
So if you call `ncdirect_bg_rgb8` and `ncdirect_fg_rgb8` before calling `ncdirect_putstr` is it going to use those colors?

**Comment by igo95862** [comment]
```
uint64_t channels = 0;
    if (channels_object != NULL)
    {
        channels = channels_object->nc_channels;
    }
    else
    {
        int foreground_colors = ncdirect_fg_default(self->ncdirect_ptr);
        int background_colors = ncdirect_bg_default(self->ncdirect_ptr);
        channels_set_bchannel(&channels, background_colors);
        channels_set_fchannel(&channels, foreground_colors);
    }
```
I've put something like this in the code.

**Comment by dankamongmen** [comment]
> So if you call `ncdirect_bg_rgb8` and `ncdirect_fg_rgb8` before calling `ncdirect_putstr` is it going to use those colors?

no, ncdirect_putstr() is always going to set the color. if you want say red on green (yuck!), you have three ways:

1) `ncdirect_fg_rgb() + ncdirect_bg_rgb() + printf()`
2) `uint64_t channels = CHANNEL_RGB_INITIALIZER(0xff, 0, 0, 0, 0xff, 0); ncdirect_putstr(channels)`
3) `uint64_t channels = 0; channels_set_fg_rgb(&channels, 0xff, 0, 0); channels_set_bg_rgb(&channels, 0, 0xff, 0); ncdirect_putstr(channels)`

**Comment by dankamongmen** [comment]
The "default color" means the default colors configured in the terminal emulator, not the "active colors". I know that isn't exactly clear :(.

**Comment by igo95862** [comment]
Ohh... 

Well I thought it behaves same way as surfaces.

**Comment by dankamongmen** [comment]
with ncplanes it's the same deal, just none of the output functions there take a channel, AFAIK.

* there's a *defaut fg* and *default bg* which are properties of the **terminal emulator due to configuration**,
* there is an *active fg* and *active bg* which are properties of a `ncplane`
* there is an *active fg* and *active bg* which are properties of the **terminal emulator due to recent escape sequences**

Telling the terminal to use color index -1 means "use the configured defaults", and this is what an all 0's channel does. By setting the `CELL_NOT_DEFAULT` flag or whatever (you don't need manage this), the RGB bits are instead used.
